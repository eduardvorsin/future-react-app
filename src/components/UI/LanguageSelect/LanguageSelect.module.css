.language__select {
	flex-wrap: wrap;
}

.language__select>label {
	font-size: 0;
}

[data-theme='light'] .language__select>div,
[data-theme='light'] .language__select select {
	background-color: var(--dark-color-100);
}

[data-theme='light'] .language__select select {
	color: var(--light-color-100);
}