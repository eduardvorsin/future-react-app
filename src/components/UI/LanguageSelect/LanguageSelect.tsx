import React, { ChangeEventHandler, FC, useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';
import Select from '../Select/Select';
import classes from './LanguageSelect.module.css';
import i18n from '../../../localization/i18next';
import { LanguageContext, Languages } from '../../../contexts/LanguageContext';

type LanguageOption = {
  value: Languages,
  label: string,
}

const languagesOptions: LanguageOption[] = [
  {
    value: 'en',
    label: 'ENG',
  },
  {
    value: 'ru',
    label: 'RUS',
  },
];

type LanguageSelect = {
  className?: string,
  onChange?: ChangeEventHandler<HTMLSelectElement>,
}

const LanguageSelect: FC<LanguageSelect> = ({
  className,
  onChange,
}) => {
  const { t } = useTranslation();
  const language = useContext(LanguageContext);
  const [selectValue, setSelectValue] = useState<string>(language.value);
  const languageChangeHandler: ChangeEventHandler<HTMLSelectElement> = (e) => {
    const currentLanguage = e.target.value as Languages;
    setSelectValue(currentLanguage);
    i18n.changeLanguage(currentLanguage);
    language.selectLanguage(currentLanguage);

    if (onChange) onChange(e);
  };

  return (
    <Select
      id='language-select'
      className={`${classes.language__select} ${className}`}
      labelText={t('languageSelection')}
      options={languagesOptions}
      value={selectValue}
      onChange={languageChangeHandler}
    />
  );
};

export default LanguageSelect;
