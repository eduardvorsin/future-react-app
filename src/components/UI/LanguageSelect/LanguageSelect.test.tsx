/* eslint-disable @typescript-eslint/no-empty-function */
import React, { ChangeEvent } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';
import LanguageSelect from './LanguageSelect';
import i18n from '../../../localization/i18next';

describe('LanguageSelect tests', () => {
  beforeAll(() => {
    jest.mock('react-i18next', () => ({
      ...jest.requireActual('react-i18next'),
      useTranslation: () => ({
        t: (str: string) => str,
        i18n: {
          changeLanguage: () => new Promise(() => { }),
        },
      }),
    }));
  });

  it('is rendered correctly', () => {
    render(<LanguageSelect />);

    expect(screen.getByRole<HTMLSelectElement>('combobox')).toBeInTheDocument();
  });

  it('when you change the select option, the mock function should be started', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: ChangeEvent]>();
    render(
      <LanguageSelect
        onChange={mockFn}
      />,
    );

    user.selectOptions(screen.getByRole<HTMLSelectElement>('combobox'), 'ru');

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalledTimes(1);
    });
  });

  it('is a basic snapshot', () => {
    render(<LanguageSelect />);

    expect(screen.getByRole<HTMLSelectElement>('combobox')).toMatchSnapshot();
  });
});

describe('LanguageSelect integration tests', () => {
  it('the component with the text in English is displayed correctly', () => {
    const localizedText = i18n.t('languageSelection');

    render(
      <I18nextProvider i18n={i18n}>
        <LanguageSelect />
      </I18nextProvider>,
    );

    expect(screen.getByLabelText<HTMLDivElement>(localizedText)).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const localizedText = i18n.t('languageSelection');

    render(
      <I18nextProvider i18n={i18n}>
        <LanguageSelect />
      </I18nextProvider>,
    );

    expect(screen.getByLabelText<HTMLDivElement>(localizedText)).toBeInTheDocument();
  });
});
