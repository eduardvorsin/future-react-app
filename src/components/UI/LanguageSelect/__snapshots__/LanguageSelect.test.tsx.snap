// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelect tests is a basic snapshot 1`] = `
<select
  class="select-wrapper__select"
  id="language-select"
>
  <option
    value="en"
  >
    ENG
  </option>
  <option
    value="ru"
  >
    RUS
  </option>
</select>
`;
